### ISSUE TITLE
IGV on OS X can only load one BAM file, all subsequent bam load actions will fail

### ISSUE THREAD

**Issue by MattBashton** [igvteam/igv#1053]
HI I've had this problem for some time and it's a usability pain, if I load a BAM into IGV either by dragging it into the window or using he File menu all is fine (for this first file), but attempting to load a second file will always fail.  As far as I can tell this is not related to JVM size as have increased it to 4GB and there is plenty of head room following initial load. 

I've started IGV from the terminal and the following error `java.awt.datatransfer.UnsupportedFlavorException: Unicode String` appears in the terminal output:

```
echo Using bundled JDK.
WARNING: package com.sun.java.swing.plaf.windows not in java.desktop
WARNING: package sun.awt.windows not in java.desktop
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.10+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.10+9, mixed mode)
WARN [2021-11-09T17:23:18,004]  [Globals.java:135] [main]  Development mode is enabled
INFO [2021-11-09T17:23:18,021]  [DirectoryManager.java:166] [main]  IGV Directory: /Users/bashton/igv
INFO [2021-11-09T17:23:18,940]  [Main.java:208] [AWT-EventQueue-0]  Startup  IGV Version 2.11.2 10/07/2021 10:41 PM
INFO [2021-11-09T17:23:18,951]  [Main.java:209] [AWT-EventQueue-0]  Java 11.0.10 (build 11.0.10+9) 2021-01-19
INFO [2021-11-09T17:23:18,952]  [Main.java:212] [AWT-EventQueue-0]  Java Vendor: AdoptOpenJDK https://adoptopenjdk.net/
INFO [2021-11-09T17:23:18,952]  [Main.java:214] [AWT-EventQueue-0]  JVM: OpenJDK 64-Bit Server VM AdoptOpenJDK   
INFO [2021-11-09T17:23:19,075]  [Main.java:217] [AWT-EventQueue-0]  Default User Directory: /Users/bashton
INFO [2021-11-09T17:23:19,076]  [Main.java:218] [AWT-EventQueue-0]  OS: Mac OS X 10.14.6 x86_64
INFO [2021-11-09T17:23:19,166]  [CommandListener.java:122] [Thread-3]  Listening on port 60151
INFO [2021-11-09T17:23:20,559]  [GenomeManager.java:183] [pool-3-thread-1]  Loading genome: /Users/bashton/igv/genomes/ASM985889v3.json
INFO [2021-11-09T17:23:20,907]  [IGV.java:1326] [pool-3-thread-1]  Loading 1 resources.
INFO [2021-11-09T17:23:20,912]  [TrackLoader.java:123] [pool-3-thread-1]  Loading resource, path https://s3.amazonaws.com/igv.org.genomes/ASM985889v3/GCF_009858895.2_ASM985889v3_genomic.gff.gz
INFO [2021-11-09T17:23:31,528]  [TrackLoader.java:123] [pool-3-thread-2]  Loading resource, path /Volumes/SD card/GATK_test/1.bam
ERROR [2021-11-09T17:23:31,530]  [DataPanelContainer.java:250] [AWT-EventQueue-0]  java.awt.datatransfer.UnsupportedFlavorException: Unicode String
INFO [2021-11-09T17:23:37,257]  [TrackLoader.java:123] [pool-3-thread-3]  Loading resource, path /Volumes/SD card/GATK_test/2.bam
ERROR [2021-11-09T17:23:37,258]  [DataPanelContainer.java:250] [AWT-EventQueue-0]  java.awt.datatransfer.UnsupportedFlavorException: Unicode String
```

I'm then left with an empty data track with the file name on the far left but no actual data is loaded and no reads display.

**Comment by davideby** [comment]
Hi,

What version of macOS are you using, and have you noticed this with multiple macOS releases?

From your stack trace error message, it looks like you are working from a removable drive (SD card).  Recent macOS releases have tightened security around applications accessing certain locations, among them removable drives.  Can you try opening the same files from somewhere on your hard drive instead, just to check whether that is the problem?

There are also new restrictions around the Downloads folder, Documents folders, and the Desktop.  None of those should be a problem, but for the sake of the experiment could you try from some other location like a temp folder in your home directory?

Thanks,
David

**Comment by MattBashton** [comment]
Hi, Yes I get identical behaviour from a temp dir located in my home directory, first bam file will load fine, second stalls, my OS is: Mac OS X 10.14.6 x86_64 I've also noticed this on 10.13 too.

**Comment by MattBashton** [comment]
Granting IGV full disk access permissions also has no effect.

**Comment by davideby** [comment]
Thanks @MattBashton.  Probably @jrobinso will need to take a closer look at this one.

**Comment by jrobinso** [comment]
This is strange, I can't reproduce it.   I've loaded up to 4 bams, haven't tried beyond that.    If this was a widespread problem I would expect more reports.  

Could you be more specific on what you mean by "second one will fail"?   Fail in what way?

**Comment by jrobinso** [comment]
I assume that is the entire log or console output?  

I also see the "UnsupportedFlavorException" when using drag&drop,  I don't think that's related.   I don't know where to go with this if I can reproduce it,  I doubt it has anything to do with your specific bams but if you could zip up a small test case with steps to reproduce I will try.

**Comment by MattBashton** [comment]
That was the console output, I launched it from the command line to see if anything interesting was there.

**Comment by MattBashton** [comment]
Okay here is an example (went with something with low coverage so smaller file size). As you can see from screenshot after loading in one of the files (1.mapped.bam), and scrolling all the way down, you can just see reads on the far right from the first file, but the second is not loading in any reads. This behaviour is universally an issue for me, I simply can't get IGV to ever load more than one file, so always have to close before opening another file.
![Screenshot 2021-11-10 at 11 50 22](https://user-images.githubusercontent.com/3043019/141108114-0b385a10-d8c7-4f24-a3af-c6033dbc2b2a.png)
[Bam.zip](https://github.com/igvteam/igv/files/7512385/Bam.zip)

**Comment by m-crown** [comment]
I also have this issue. 

```
echo Using bundled JDK.
WARNING: package com.sun.java.swing.plaf.windows not in java.desktop
WARNING: package sun.awt.windows not in java.desktop
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.10+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.10+9, mixed mode)
WARN [2021-11-10T11:50:02,696]  [Globals.java:135] [main]  Development mode is enabled
INFO [2021-11-10T11:50:02,728]  [DirectoryManager.java:166] [main]  IGV Directory: /Users/matthewcrown/igv
INFO [2021-11-10T11:50:04,334]  [Main.java:208] [AWT-EventQueue-0]  Startup  IGV Version 2.11.2 10/07/2021 10:41 PM
INFO [2021-11-10T11:50:04,344]  [Main.java:209] [AWT-EventQueue-0]  Java 11.0.10 (build 11.0.10+9) 2021-01-19
INFO [2021-11-10T11:50:04,345]  [Main.java:212] [AWT-EventQueue-0]  Java Vendor: AdoptOpenJDK https://adoptopenjdk.net/
INFO [2021-11-10T11:50:04,346]  [Main.java:214] [AWT-EventQueue-0]  JVM: OpenJDK 64-Bit Server VM AdoptOpenJDK   
INFO [2021-11-10T11:50:04,682]  [Main.java:217] [AWT-EventQueue-0]  Default User Directory: /Users/matthewcrown
INFO [2021-11-10T11:50:04,683]  [Main.java:218] [AWT-EventQueue-0]  OS: Mac OS X 10.16 x86_64
INFO [2021-11-10T11:50:04,856]  [CommandListener.java:122] [Thread-3]  Listening on port 60151
INFO [2021-11-10T11:50:07,393]  [GenomeManager.java:183] [pool-3-thread-1]  Loading genome: /Users/matthewcrown/igv/genomes/ASM985889v3.json
INFO [2021-11-10T11:50:07,821]  [IGV.java:1326] [pool-3-thread-1]  Loading 1 resources.
INFO [2021-11-10T11:50:07,834]  [TrackLoader.java:123] [pool-3-thread-1]  Loading resource, path https://s3.amazonaws.com/igv.org.genomes/ASM985889v3/GCF_009858895.2_ASM985889v3_genomic.gff.gz
INFO [2021-11-10T11:51:34,765]  [IGV.java:1326] [pool-3-thread-2]  Loading 2 resources.
INFO [2021-11-10T11:51:34,768]  [TrackLoader.java:123] [pool-3-thread-2]  Loading resource, path /Users/matthewcrown/trimPrimerSequences/test1.mapped.primertrimmed.sorted.bam
INFO [2021-11-10T11:51:34,990]  [TrackLoader.java:123] [pool-3-thread-2]  Loading resource, path /Users/matthewcrown/trimPrimerSequences/test2.mapped.primertrimmed.sorted.bam
```
Using  macOS 11.6 (20G165)

**Comment by jrobinso** [comment]
Interesting, @m-crown you are using the same assembly (Sars-Cov-2).   This issue seems to be specific to that assembly, have either of you seen it on any other assembly (such as human)?  

I can reproduce it with these files and am looking into it.

**Comment by jrobinso** [comment]
The puzzle I'm working on at the moment is why you are ever seeing any alignments at all.   The only sequence defined in the reference you are using is. ```NC_045512.2```.   However your reads are aligned to sequence ```MN908947.3```.  So actually you should never see any reads from either file regardless of the order they are loaded.   Changing the name of your sequence in your bam files to match the reference fixes the issue (fixed files attached).

In summary there is a bug here, but not the one we thought.  You should have not seen any alignments on the first (or second) file load, but should have received a warning to the effect that no reads were found matching any sequences in your reference.

@m-crown You could be seeing the same issue, check the sequence name(s) in your bam

```
samtools view -H your.bam
```
[fixed.zip](https://github.com/igvteam/igv/files/7514633/fixed.zip)

**Comment by m-crown** [comment]
NC_045512.2 and MN908947.3 are the same reference sequence, just different identifiers. Using the MN908947.3 fasta allows both files to be loaded.

**Comment by jrobinso** [comment]
Understood, but IGV of course doesn't magically know that.  Its still a mystery why alignments were shown against the "wrong" sequence (from the point of view of the program), I will need to get to the bottom of that.    However thanks for the  information on the identifiers,  I can alias them in the genome definition.   If there are any other identifiers for this sequence post them here,  I don't not work on this assembly personally.

**Comment by m-crown** [comment]
Yep, I think that would have been more obvious to me had no alignments been shown for a single sample - definitely a weird behaviour to show for one but not multiple, but at least with the correct identifier things work.

**Comment by jrobinso** [comment]
@m-crown Yes this is a very misleading behavior, apologies, I will fix it.   

I have considered the idea of aliasing sequence by length, in other words if a name match isn't found use the length of the sequence.   The idea being that given the user has chosen a reference, the likelihood that an exact match on length would occur by chance is small.

**Comment by MattBashton** [comment]
Thanks @jrobinso this now works fine. I was aware of the genome assembly aliasing feature in IGV and I had (incorrectly) assumed this must be working for (NC_045512.2/MN908947.3) otherwise I'd not see the first bam file I loaded, I guess I was wrong, but now I'm more puzzled I got the bam to display in the first instance!

**Comment by jrobinso** [comment]
@MattBashton I'm puzzled as well!  I'm in meetings now but am going to dig into this later.  When its fixed I will rename this issue to reflect whatever the real problem was and close it.

**Comment by jrobinso** [comment]
OK mystery solved.  I had actually implemented matching sequences by length if no name match is found some time ago, and had forgotten it.   However there was a bug that resulted in this matching algorithm working on only 1 file,  some state was remembered that caused it to skip the check on subsequent loads.   This bug would only affect cases where the bam file sequence name and reference sequence name mismatched unexpectedly (no aliases defined), and furthermore would not be seen until the second such file was loaded.  This explains the lack of reports, well not total lack,  I'm really glad you reported it.
